### DISCUSSION TITLE
Sampling uniformly from entities in a query

### DISCUSSION THREAD

**Discussion by CharlesJQuarra** [SanderMertens/flecs#664] [Q&A]
Hi, forgive me if the question is too basic or is already explained somewhere. I've been searching for this on the queries documentation and I couldn't find the answer

Basically I want O(k) access (let's pretend page faults are irrelevant/not a bottleneck for this) to an uniformly sampled batch of k entities (with replacement, sampling without replacement is harder) that should be accessible from a given query.

Right now the only way I've found to implement this is to loop at least once on a query with `each`, store the entity ids on a std::vector, and sample from there. Let me know if there is a way to avoid this extra setup (or at the very least, know the size of the query set beforehand, so I can reserve enough memory when initializing the `std::vector` )

**Comment by ZeroErrors** [comment]
Hi @CharlesJQuarra

Did you find an answer to your question?

If not can you provide a bit more information about what you are trying to do.

My guess based on your description of storing the entity id's in a vector is you should use the `iter` function instead of `each`, this operates on each table instead of an individual entity.
```cpp
f.iter([](flecs::iter& it) {
    for (size_t i = 0; i < it.count(); i++) {
      flecs::entity e = it.entity(i);
      // Do stuff with the entity
    }
});
```
